### PR TITLE
Scaladocs added

### PR DESCRIPTION
**Pull request by salamonpavel**
Missing scaladocs added

### CODE REVIEW

File: agent/src/main/scala/za/co/absa/atum/agent/AtumContext.scala [hunk 1]
```diff
@@ -54,6 +64,17 @@ class AtumContext private[agent] (
     }
   }
 
+  /**
+   * Creates a checkpoint in the AtumContext.
+   *
+   * This method is used to mark a specific point in the data processing pipeline where measurements of data quality are taken.
```
Comment (benedeki) [inline agent/src/main/scala/za/co/absa/atum/agent/AtumContext.scala:70]:
```suggestion
   * This method is used to mark a specific point in the data processing pipeline where measurements of data completeness are taken.
```

File: agent/src/main/scala/za/co/absa/atum/agent/AtumContext.scala [hunk 2]
```diff
@@ -54,6 +64,17 @@ class AtumContext private[agent] (
...
+   * The checkpoint is identified by a unique name, which can be used later to retrieve the measurements taken at this point.
```
Comment (benedeki) [inline agent/src/main/scala/za/co/absa/atum/agent/AtumContext.scala:71]: The name is not necessary unique. Not even within the same partitioning.

File: agent/src/main/scala/za/co/absa/atum/agent/dispatcher/Dispatcher.scala [hunk 3]
```diff
@@ -18,8 +18,20 @@ package za.co.absa.atum.agent.dispatcher
 
 import za.co.absa.atum.model.dto.{AtumContextDTO, CheckpointDTO, PartitioningSubmitDTO}
 
+/**
+ *  This trait provides a contract for different dispatchers
+ */
 trait Dispatcher {
+  /**
+   *  This method is used to fetch AtumContext from server.
```
Comment (benedeki) [inline agent/src/main/scala/za/co/absa/atum/agent/dispatcher/Dispatcher.scala:26]:
```suggestion
   *  This method is used to ensure the server knows the given partitioning.. As  response the `AtumContext` is fetched from the server.
```

File: agent/src/main/scala/za/co/absa/atum/agent/dispatcher/Dispatcher.scala [hunk 4]
```diff
@@ -18,8 +18,20 @@ package za.co.absa.atum.agent.dispatcher
...
+   *  @param partitioning: PartitioningSubmitDTO to be used for fetching AtumContext.
```
Comment (benedeki) [inline agent/src/main/scala/za/co/absa/atum/agent/dispatcher/Dispatcher.scala:27]: If above accepted, this should be reworded too, IMHO.

File: agent/src/main/scala/za/co/absa/atum/agent/exception/AtumAgentException.scala [hunk 5]
```diff
@@ -16,15 +16,35 @@
...
 
+/**
+ * This class represents an exception thrown by the Atum Agent.
+ *
+ * @param message A message describing the exception.
+ */
 abstract class AtumAgentException(message: String) extends Exception(message)
 
+/**
+ * This object contains all the possible exceptions thrown by the Atum Agent.
```
Comment (benedeki) [inline agent/src/main/scala/za/co/absa/atum/agent/exception/AtumAgentException.scala:27]: Not sure if necessary _all_.

File: agent/src/main/scala/za/co/absa/atum/agent/AtumContext.scala (hunk 1 above)
Comment (salamonpavel) [reply agent/src/main/scala/za/co/absa/atum/agent/AtumContext.scala:70]: Changed as per your suggestion.

File: agent/src/main/scala/za/co/absa/atum/agent/dispatcher/Dispatcher.scala (hunk 3 above)
Comment (salamonpavel) [reply agent/src/main/scala/za/co/absa/atum/agent/dispatcher/Dispatcher.scala:26]: Changed as per your suggestion.

File: agent/src/main/scala/za/co/absa/atum/agent/dispatcher/Dispatcher.scala (hunk 4 above)
Comment (salamonpavel) [reply agent/src/main/scala/za/co/absa/atum/agent/dispatcher/Dispatcher.scala:27]: Changed as per your suggestion.

File: agent/src/main/scala/za/co/absa/atum/agent/exception/AtumAgentException.scala (hunk 5 above)
Comment (salamonpavel) [reply agent/src/main/scala/za/co/absa/atum/agent/exception/AtumAgentException.scala:27]: Changed as per your suggestion.

File: agent/src/main/scala/za/co/absa/atum/agent/AtumAgent.scala [hunk 6]
```diff
@@ -76,6 +76,13 @@ class AtumAgent private[agent] () {
     getExistingOrNewContext(atumPartitions, atumContext)
   }
 
+  /**
+   * Provides an AtumContext given a `AtumPartitions` instance for sub partitions.
+   * Retrieves the data from AtumService API.
+   * @param subPartitions
+   * @param parentAtumContext
```
Comment (lsulak) [inline agent/src/main/scala/za/co/absa/atum/agent/AtumAgent.scala:83]: todo document? For the sake of completeness, since it's semi-auto-generated

File: agent/src/main/scala/za/co/absa/atum/agent/AtumContext.scala [hunk 7]
```diff
@@ -54,6 +64,17 @@ class AtumContext private[agent] (
     }
   }
 
+  /**
+   * Creates a checkpoint in the AtumContext.
+   *
+   * This method is used to mark a specific point in the data processing pipeline where measurements of data completeness are taken.
+   * The checkpoint is identified by a name, which can be used later to retrieve the measurements taken at this point.
```
Comment (lsulak) [inline agent/src/main/scala/za/co/absa/atum/agent/AtumContext.scala:71]: some of these lines are longer than 120 characters

File: agent/src/main/scala/za/co/absa/atum/agent/AtumContext.scala [hunk 8]
```diff
@@ -54,6 +64,17 @@ class AtumContext private[agent] (
     }
   }
 
+  /**
+   * Creates a checkpoint in the AtumContext.
+   *
+   * This method is used to mark a specific point in the data processing pipeline where measurements of data completeness are taken.
```
Comment (lsulak) [inline agent/src/main/scala/za/co/absa/atum/agent/AtumContext.scala:70]: "of data completeness" - not sure if we want to be this specific here 🤔

File: agent/src/main/scala/za/co/absa/atum/agent/AtumContext.scala [hunk 9]
```diff
@@ -91,24 +119,50 @@ class AtumContext private[agent] (
...
+   */
   def addMeasure(newMeasure: Measure): AtumContext = {
     measures = measures + newMeasure
     this
   }
 
+  /**
+   * Adds multiple measures to the AtumContext.
+   *
+   * @param measures the sequence of measures to be added
```
Comment (lsulak) [inline agent/src/main/scala/za/co/absa/atum/agent/AtumContext.scala:154]:
```suggestion
   * @param measures set of measures to be added
```

probably just for better accuracy

File: agent/src/main/scala/za/co/absa/atum/agent/dispatcher/Dispatcher.scala [hunk 10]
```diff
@@ -18,8 +18,21 @@ package za.co.absa.atum.agent.dispatcher
...
+  /**
+   *  This method is used to ensure the server knows the given partitioning. 
+   *  As a response the `AtumContext` is fetched from the server.
+   *  @param partitioning: PartitioningSubmitDTO to be used to ensure server knows the given partitioning.
+   *  @return AtumContextDTO.
+   */
   def createPartitioning(partitioning: PartitioningSubmitDTO): AtumContextDTO
 
+  /**
+   *  This method is used to save checkpoint to server.
```
Comment (lsulak) [inline agent/src/main/scala/za/co/absa/atum/agent/dispatcher/Dispatcher.scala:34]: or, to data store? Or, to send to server sfor storing it into data store? Or we don't want to be that precise, since it's a general dispatcher?

File: agent/src/main/scala/za/co/absa/atum/agent/exception/AtumAgentException.scala [hunk 11]
```diff
@@ -16,15 +16,35 @@
 
 package za.co.absa.atum.agent.exception
 
+/**
+ * This class represents an exception thrown by the Atum Agent.
+ *
+ * @param message A message describing the exception.
```
Comment (lsulak) [inline agent/src/main/scala/za/co/absa/atum/agent/exception/AtumAgentException.scala:22]: I would even remove this line completely from all occurrences, not really bringing any value

File: agent/src/main/scala/za/co/absa/atum/agent/exception/AtumAgentException.scala [hunk 12]
```diff
@@ -16,15 +16,35 @@
 
 package za.co.absa.atum.agent.exception
 
+/**
+ * This class represents an exception thrown by the Atum Agent.
```
Comment (lsulak) [inline agent/src/main/scala/za/co/absa/atum/agent/exception/AtumAgentException.scala:20]:
```suggestion
 * This class represents a basis for exceptions thrown by the Atum Agent.
```
?

Comment (lsulak) [reply agent/src/main/scala/za/co/absa/atum/agent/exception/AtumAgentException.scala:20]: or something similar

File: agent/src/main/scala/za/co/absa/atum/agent/model/Measure.scala [hunk 13]
```diff
@@ -157,6 +163,15 @@ object Measure {
     override val resultValueType: ResultValueType.ResultValueType = ResultValueType.String
   }
 
+  /**
+   * This method aggregates a column of a given data frame using a given aggregation expression.
+   * The result is converted to a string.
+   *
+   * @param ds            A data frame
```
Comment (lsulak) [inline agent/src/main/scala/za/co/absa/atum/agent/model/Measure.scala:170]:
```suggestion
   * @param df            A data frame
``` 

?

File: agent/src/main/scala/za/co/absa/atum/agent/model/MeasuresBuilder.scala [hunk 14]
```diff
@@ -20,6 +20,9 @@ import za.co.absa.atum.agent.exception.AtumAgentException.MeasureException
 import za.co.absa.atum.agent.model.Measure._
 import za.co.absa.atum.model.dto
 
+/**
+ * This object provides a functionality to convert a DTO representation of measures to their internal representation.
```
Comment (lsulak) [inline agent/src/main/scala/za/co/absa/atum/agent/model/MeasuresBuilder.scala:24]:
```suggestion
 * This object provides a functionality to convert a DTO representation of measures to the Agent's internal representation of this object.
```

or something similar?

File: agent/src/main/scala/za/co/absa/atum/agent/dispatcher/Dispatcher.scala (hunk 10 above)
Comment (lsulak) [reply agent/src/main/scala/za/co/absa/atum/agent/dispatcher/Dispatcher.scala:34]: Actually no, not all dispatchers are/will be working with data stores

File: agent/src/main/scala/za/co/absa/atum/agent/dispatcher/Dispatcher.scala [hunk 15]
```diff
@@ -18,8 +18,21 @@ package za.co.absa.atum.agent.dispatcher
 
 import za.co.absa.atum.model.dto.{AtumContextDTO, CheckpointDTO, PartitioningSubmitDTO}
 
+/**
+ *  This trait provides a contract for different dispatchers
+ */
 trait Dispatcher {
+  /**
+   *  This method is used to ensure the server knows the given partitioning.
```
[the remark this answers is not in the document]
Comment (lsulak) [reply agent/src/main/scala/za/co/absa/atum/agent/dispatcher/Dispatcher.scala:26]: Actually no, it's okay. Data store is too specific and not all dispatchers will work with it

[the remark this answers is not in the document]
Comment (lsulak) [reply agent/src/main/scala/za/co/absa/atum/agent/dispatcher/Dispatcher.scala:26]: But I would probably use similar logic here as for the checkpoint creation, so something like `save` or `store` instead of `knows`

File: agent/src/main/scala/za/co/absa/atum/agent/AtumContext.scala (hunk 8 above)
Comment (salamonpavel) [reply agent/src/main/scala/za/co/absa/atum/agent/AtumContext.scala:70]: I think it's reasonable. But if we find it too wordy/detailed/specific we can alter the doc string in the future. Let's keep it for now.

File: agent/src/main/scala/za/co/absa/atum/agent/dispatcher/Dispatcher.scala (hunk 15 above)
[the remark this answers is not in the document]
Comment (salamonpavel) [reply agent/src/main/scala/za/co/absa/atum/agent/dispatcher/Dispatcher.scala:26]: Not a significant difference in my opinion. As dispatcher communicates with serves I would keep server, even though at the end of the day it's the database where the data are stored. If you feel strongly about this I will change it. @lsulak

File: agent/src/main/scala/za/co/absa/atum/agent/dispatcher/Dispatcher.scala (hunk 10 above)
Comment (salamonpavel) [reply agent/src/main/scala/za/co/absa/atum/agent/dispatcher/Dispatcher.scala:34]: I guess this depends on how we resolve the one above.

File: agent/src/main/scala/za/co/absa/atum/agent/exception/AtumAgentException.scala (hunk 12 above)
Comment (salamonpavel) [reply agent/src/main/scala/za/co/absa/atum/agent/exception/AtumAgentException.scala:20]: Change to "This type represents a base class for exceptions thrown by the Atum Agent." which is close to your suggestion emphasizing it's a basis for other exceptions.

File: agent/src/main/scala/za/co/absa/atum/agent/model/MeasuresBuilder.scala (hunk 14 above)
Comment (salamonpavel) [reply agent/src/main/scala/za/co/absa/atum/agent/model/MeasuresBuilder.scala:24]: Don't see much of a difference. Let me know if you feel strongly about this, otherwise I suggest to keep it as it is.

File: agent/src/main/scala/za/co/absa/atum/agent/dispatcher/Dispatcher.scala (hunk 15 above)
[the remark this answers is not in the document]
Comment (salamonpavel) [reply agent/src/main/scala/za/co/absa/atum/agent/dispatcher/Dispatcher.scala:26]: Btw the "server knows" was David's suggestion.

File: agent/src/main/scala/za/co/absa/atum/agent/dispatcher/Dispatcher.scala (hunk 10 above)
Comment (lsulak) [reply agent/src/main/scala/za/co/absa/atum/agent/dispatcher/Dispatcher.scala:34]: yeah let's keep it as is then

File: agent/src/main/scala/za/co/absa/atum/agent/exception/AtumAgentException.scala (hunk 11 above)
Comment (lsulak) [reply agent/src/main/scala/za/co/absa/atum/agent/exception/AtumAgentException.scala:22]: still todo potentially?

File: agent/src/main/scala/za/co/absa/atum/agent/model/MeasuresBuilder.scala (hunk 14 above)
Comment (lsulak) [reply agent/src/main/scala/za/co/absa/atum/agent/model/MeasuresBuilder.scala:24]: I think my suggestion makes it easier to read for someone not familiar with the codebase & it's more exact

File: agent/src/main/scala/za/co/absa/atum/agent/exception/AtumAgentException.scala (hunk 11 above)
Comment (salamonpavel) [reply agent/src/main/scala/za/co/absa/atum/agent/exception/AtumAgentException.scala:22]: It's a standard part of java/scala doc for a class with constructor parameters. I would rather keep it. Some exceptions might have different parameters (HttpException for instance).

File: agent/src/main/scala/za/co/absa/atum/agent/model/MeasuresBuilder.scala (hunk 14 above)
Comment (salamonpavel) [reply agent/src/main/scala/za/co/absa/atum/agent/model/MeasuresBuilder.scala:24]: Ok. Changed.
